core/src/transaction.rs

```diff
@@ -177,8 +181,8 @@ impl Transaction {
         self.inputs.len() == 1 && self.inputs[0].previous_output.is_null()
     }
 
-    pub fn hash(&self) -> H256 {
-        sha3_256(serialize(self).unwrap()).into()
+    pub fn hash(&self) -> &H256 {
```

Comment:
We need to clone this ref in most cases, can we change it to returning owned copy?

Reply:
It's meant to be, you should use borrow as you can, 
If you need a copy, explicitly clone it.

The reason for `Hash` and `Uint` types no longer implement `Copy` is prevent unconscious copy. Essentially, `Hash` is an array filled with byte,  when we use the array, we prefer to using `slice` pointer to access it instead of copy everywhere.

Most other languages copy and allocate all the time by default, while Rust gives a choice of avoiding that in several ways. the same thought here.